### DISCUSSION TITLE
아이템 36. 비트 필드 대신 EnumSet을 사용하라.

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Effective-Java/Effective-Java#91] [3. 과제] [answered]
📝 9️⃣성

- [0. 들어가기에 앞서 🧐](#0------------)
- [1. 비트 필드 ✨](#1--------)
  * [1.1 비트 필드 사용](#11---------)
  * [1.2 비트 필드의 문제점](#12-----------)
    + [1. 제한된 확장성](#1--------)
    + [2. 부적절한 캡슐화](#2---------)
- [2. 비트 필드 대신 EnumSet을 사용하라 ⭐️](#2----------enumset---------)
  * [2.1 EnumSet 이란?](#21-enumset----)
  * [2.2 EnumSet 사용시 고려할 사항](#22-enumset-----------)
  * [2.3 EnumSet의 구현체](#23-enumset-----)
  * [2.4 비트 필드 컬러 클래스를 EnumSet으로 리팩토링 해보자 💎](#24---------------enumset--------------)
- [3. 핵심 정리 📚](#3---------)
- [4. 회고 🧹](#4------)

<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>

---

# 0. 들어가기에 앞서 🧐

- [비트 연산자](https://coding-factory.tistory.com/521)에 대해 한번 읽어본 뒤 접근하자. (가볍게)

# 1. 비트 필드 ✨

비트 필드는 컴퓨터 과학의 역사 초기에 개발된 `메모리 절약 도구`이다.
여러 true - false 옵션을 `하나의 바이트`로 결합하여 메모리를 절약한다.
true - false 옵션은 각 옵션에 대해 전체 바이트를 사용하는 대신 비트 필드로 `그룹화(집합)` 할 수 있다.

## 1.1 비트 필드 사용

비트 필드를 사용하면 비트별 연산을 사용해 합집합과 교집합 같은 집합 연산을 효율적으로 수행할 수 있다.

- `빛의 삼원색` 예제로 알아 보자.

<details>
<summary>Color 클래스</summary>
<div markdown="1">

```java

package item36;

public class Color {
    
    public static final int RED = 1; // 1 << 0
    public static final int GREEN = 2; // 1 << 1
    public static final int BLUE = 4; // 1 << 2

    private int colors;

    public int getColors() {
        return colors;
    }

    public void setColors(int colors) {
        this.colors = colors;
    }
}

```

</div>
</details>

<details>
<summary>데모 클래스</summary>
<div markdown = "1">

```java

package item36;

public class DemoBitFieldsColor {

    private Color demoColor;
    
    public void setBitFields() {
        this.demoColor = new Color();
        this.demoColor.setColors(Color.RED | Color.GREEN);
    }

    public void printColors() {
        if (demoColor.getColors() == (Color.RED | Color.GREEN)) {
            System.out.println("YELLOW");
        }

        System.out.println("-- 포함된 색상 --");

        if ((demoColor.getColors() & Color.RED) > 0) {
            System.out.println("RED");
        }

        if ((demoColor.getColors() & Color.BLUE) > 0) {
            System.out.println("BLUE");
        }
    }
}


```
</div>
</details>

<details>
<summary>코드 실행 결과</summary>
<div>

<img width="761" alt="image" src="https://user-images.githubusercontent.com/53285909/215102389-bf0c9bca-0ee4-4585-aca5-3ed9a6120a3d.png">

</div>
</details>

> 자바에서 사용되고 있는곳은 있을까?? 🤔

<details>
<summary>Modifier(수정자)</summary>
<div markdown=“1”>

<img width="340" alt="image" src="https://user-images.githubusercontent.com/53285909/215109343-9d886db5-47b3-43c6-95ba-05e8bb66c1bc.png">

<img width="742" alt="image" src="https://user-images.githubusercontent.com/53285909/215109075-288b606e-92f6-4bdb-8dea-677d82007da9.png">

<img width="543" alt="image" src="https://user-images.githubusercontent.com/53285909/215109189-8cafb3ff-df7b-497b-8173-78732fef31b3.png">

</div>
</details>

### 1. 제한된 확장성

- 빛의 삼원색을 다시 되짚어보자

어느날 갑자기 `새로운 빛`이 추가 되었다.  해당 빛은 3의 값을 가진다.
해당 빛이 추가되면 위의 코드에서 이루어 졌던 비트 연산은 무너진다.

위의 예제는 짧았지만, 만약 수 많은 메서드가 있었고 비트 연산 값 3을 이용하고 있는 메서드가 많았다면? 

그렇지 않기를 기대 하겠지만, 인생은 항상 굴곡지다. 
개발자는 통곡한다.
애초에 비트 필드를 사용하지 않았다면...

### 2. 부적절한 캡슐화 

- 객체 지향 엔지니어링은 `캡슐화` 개념을 기반으로 한다. 즉, 객체에 있는 모든 내용은 클래스 사용자에게 숨겨야 한다. 
- 비트 필드는 사용자에게 객체의 내부 구현을 노출(Getter, Setter 등등) 하기 때문에 잘못된 캡슐화이다. 또한 여러 속성을 하나로 결합하여 인터페이스를 혼란스럽게 만든다.

다시는 비트 필드를 쓰지 않길 다짐하며 🪓 
개발자는 대안을 찾게 된다.

---

# 2. 비트 필드 대신 EnumSet을 사용하라 ⭐️

- EnumSet 클래스는 열거 타입 상수 값으로 구성된 집합을 효과적으로 표현 가능하게 한다.

## 2.1 EnumSet 이란?

- EnumSet은 enum 클래스로 작동하기 위해 특화된 Set 컬렉션이라고 할 수 있다.
- EnumSet은 Set 인터페이스를 구현하고, AbstractSet을 상속한다.
- EnumSet의 내부는 `비트 벡터`로 구현되었다.
- `비트 플래그` 대신 사용할 수 있다.

## 2.2 EnumSet 사용시 고려할 사항

1. 열거형 값만 포함 할 수 있다. 그리고 모든 값은 동일한 열거형이어야 한다.
2. null을 추가할 수 없다.
3. 스레드에 안전하지 않다. 필요한 경우 외부에서 동기화한다.
4. 복사본에 fail-safe-iterator(장애 발생시 작업을 중단하지 않는다.) 를 사용하여 컬렉션을 순회할 때, 컬렉션이 수정 되어도 `ConcurrentModificationException`이 발생하지 않는다.

## 2.3 EnumSet의 구현체

EnumSet은 추상 클래스이다. 
JDK에서는 `RegularEnumSet`, `JumboEnumSet` 2가지의 EnumSet 구현체를 제공한다.

- `RegularEnumSet`은 비트 벡터를 표현하기 위해 `단일 long 자료형`을 사용한다.
- long의 각 비트는 enum 값을 나타낸다.
- long은 64 비트의 데이터이기 때문에 64개의 원소를 저장할 수 있다.

- `JumboEnumSet`은 `long 요소의 배열`을 비트 벡터로 사용한다.
- 64개 이상의 원소를 저장한다.
- 저장된 배열 인덱스를 찾기 위해 몇 가지 추가 연산을 수행한다.

## 2.4 비트 필드 컬러 클래스를 EnumSet으로 리팩토링 해보자 💎

<details>
<summary>리팩토링 된 Color 클래스</summary>
<div markdown=“1”>

```java
package item36.ver2;

public enum Color {
    RED,
    GREEN,
    BLUE
}

```

</div>
</details>

<details>
<summary>리팩토링 된 데모 클래스</summary>
<div markdown=“1”>

```java
package item36.ver2;

import java.util.EnumSet;
import java.util.Set;

public class DemoEnumSetColor {
    
    private EnumSet<Color> demoColor;

    public void setEnumSetColor(Set<Color> color) {
        this.demoColor = EnumSet.of(Color.RED, Color.GREEN);
    }

    public void printColors() {
        if (demoColor.contains(Color.RED) && demoColor.contains(Color.GREEN)) {
            System.out.println("YELLOW");
        }

        System.out.println("-- 포함된 색상 --");

        if (demoColor.contains(Color.RED)) {
            System.out.println("RED");
        }

        if (demoColor.contains(Color.BLUE)) {
            System.out.println("BLUE");
        }
    }
}

```

</div>
</details>

실행결과는 같다.

- 이제는 `새로운 빛` 이 추가 된다 해도 그냥 `새로운 빛`을 `Color` 클래스에 추가하면 된다. 
- `데모 클래스`의 메서드가 보다 더 직관적으로 알기 쉬워졌다. 
- `Set` 인터페이스를 생성자로 받게 되기 때문에 확장성이 더 좋아졌다.

# 3. 핵심 정리 📚

- 열거할 수 있는 타입을 한데 모아 집합 형태로 사용한다고 해도 비트 필드를 사용할 이유는 없다.
- EnumSet 클래스가 이를 대체하고 있다.
- 유일한 단점은 불변 EnumSet을 구현함에 있어 `Collections.unmodifiableSet(어떤 EnumSet)`을 이용 해야한다.

---

# 4. 회고 🧹

2023.01.28 (토)

- 비트 연산이 개발하는 입장에서 얼마나 귀찮고 불편한 것을 주제로 글을 써 보려한 이유는 글보다는 코드가 나을 것이라는 생각이였다. 하지만 예제가 조금은 아쉬웠던 것 같다.
- 책의 저자 Josha Bloch가 EnumSet을 만들었다. 그래서 `사용하라` 라고 단정 짓는것은 아닌지 의심해 보려 하다가도 리팩토링 예제를 보면 고개가 끄덕여 진다.

---

> 참조
> http://underpop.online.fr/j/java/hardcore-java/hardcorejv-chp-7-sect-2.html
> https://coding-factory.tistory.com/521
> https://scshim.tistory.com/253

**Comment by JoisFe** [comment, marked as answer]
빛의 색 3개로 표현할 수 있는 케이스가 2^3 = 8가지 인데 모든 케이스를 다 저장하기 보단
빨, 초, 파를 비트 필드로 두고 OR 연산을 통해 여러 케이스를 만들 수 있게 되네요 ㅎㅎ

비트 필드는 아두이노를 이용해 LED 빛을 표현할때 많이 쓰곤 했던것 같습니다.
1, 2, 3번 LED가 있을때 1만 키고 싶을때, 1, 3을 키고 싶을 때, 1, 2, 3 모두 키고 싶을때가 다양한데
1번 LED : 첫 번째 비트
2번 LED : 두 번째 비트
3번 LED : 세 번째 비트 
로 두고 0인 경우 off, 1인 경우 on 으로 하여 3 비트로 모든 케이스를 나타낼 수 있게 되는데 위 예제가 이와 비슷한 것 같습니다.

웹 애플리케이션을 구현할 때도 특정 옵션을 선택하는 경우 선택하지 않은 경우가 있을 것이고
이러한 옵션이 매우 많다면 여러 케이스를 관리하기가 힘들텐데
비트 필드를 이용한다면 조금더 편리해질 것 같고
좀 더 나아가서 이번 주제에 맞게 비트 필드 대신 EnumSet을 이용한다면 훨씬 관리하기 좋은 코드가 될 것 같습니다.!!

**Reply by Irisation23** [reply]
확실히 `EnumSet`을 이용하게 된다면 비트 필드보다 코드량은 늘어나지만 유지보수에 유용한 코드가 되는것 같더라고요.
단순히 on / off 정도로만 이루어 지거나 true / false 정도로 이루어지는 bool 타입의 필드들에도 사용 해 볼만한 가치가 있다는 생각이 들었습니다.

**Comment by jinan159** [comment]
리플렉션에서 modifier 다룰때 or 연산 많이 사용했었는데.. EnumSet 사용하면 이렇게 직관적인 코드로 바뀌는군요
리펙토링 전 코드는 Color 내부의 값을 알아야하기 때문에 캡슐화가 깨진 코드인데, 
리펙토링 후 코드는 내부 값을 몰라도되고 확장에도 유연하다는게 아주 큰 장점이네요.
나중에 enum 을 활용할때 꼭 사용해봐야겠습니다!

**Reply by JoisFe** [reply]
글을 보니 자바의 modifier 구현 또한 비트 필드를 사용했다는 점도 처음 알았고
@jinan159 님 께서는 리플렉션에서 modifier를 다뤄보기 까지..
부족함을 많이 느낍니다.

아래 한글이나 word 사용시 특정 글에 대해 1. bold, 2. italic, 3. underline 특성을 적용할 수 있는데
이러한 경우도 비트 필드를 활용한 예시가 있고
OR 연산을 통해 각 특성을 조합하더군요 
많은 예시가 있는 것을 봤지만 특정한 비트 값을 하드코딩으로 둔다는게 맞는건가?? 라는 찝찝함이 있었는데
EnumSet을 이용한다면 이러한 찝찝함을 해결해주는 것 같습니다.

**Reply by jinan159** [reply]
@JoisFe 오.. 아래 한글 코드를 까보실 기회가 있으셨던건가요..!
신기하네요🤩

**Reply by JoisFe** [reply]
> @JoisFe 오.. 아래 한글 코드를 까보실 기회가 있으셨던건가요..! 신기하네요🤩

아니요 ㅎㅎ 구글링 통해 누가 예시를 들었더군요
해당 회사의 구체적인 코드는 모르겠습니다 ㅎㅎ
오해가 있을 수 있을 것 같네요

<img width="522" alt="image" src="https://user-images.githubusercontent.com/90208100/215311796-6cb8861a-2339-43af-9242-10428c5dd29e.png">

아래한글 공개 문서파일인데 글꼴 속성들에 대한 부분이 특정 비트위치로 나타내는 것을 보니 비트 필드와 유사한 방식으로 할 것 같은데 내부적으로 어떨지는 모르겠네요 ㅎㅎ

**Reply by Irisation23** [reply]
@jinan159 블로그의 내용도 있지만, 책에서 해당 아이템을 다루는 내용은 오히려 글꼴을 통해 다뤘지만, 제가 임의로 예제를 하나 만들어 보려고 했습니다.
다시 보게되니 오히려 더 직관적인건 글꼴이 괜찮은 것 같군요 😢
